test/cpp/microbenchmarks/callback_test_service.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "test/cpp/microbenchmarks/callback_test_service.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+grpc::string ToString(const grpc::string_ref& r) {
+  return grpc::string(r.data(), r.size());
+}
+
+int GetIntValueFromMetadataHelper(
+    const char* key,
+    const std::multimap<grpc::string_ref, grpc::string_ref>& metadata,
+    int default_value) {
+  if (metadata.find(key) != metadata.end()) {
+    std::istringstream iss(ToString(metadata.find(key)->second));
+    iss >> default_value;
+  }
+
+  return default_value;
+}
+
+int GetIntValueFromMetadata(
+    const char* key,
+    const std::multimap<grpc::string_ref, grpc::string_ref>& metadata,
+    int default_value) {
+  return GetIntValueFromMetadataHelper(key, metadata, default_value);
+}
+}  // namespace
+
+void CallbackStreamingTestService::Echo(
+    ServerContext* context, const EchoRequest* request, EchoResponse* response,
+    experimental::ServerCallbackRpcController* controller) {
+  int response_msgs_size = GetIntValueFromMetadata(
+      kServerMessageSize, context->client_metadata(), 0);
+  if (response_msgs_size > 0) {
+    response->set_message(std::string(response_msgs_size, 'a'));
+  } else {
+    response->set_message("");
+  }
+  controller->Finish(Status::OK);
+}
+
+experimental::ServerBidiReactor<EchoRequest, EchoResponse>*
+CallbackStreamingTestService::BidiStream() {
+  class Reactor
+      : public experimental::ServerBidiReactor<EchoRequest, EchoResponse> {
+   public:
+    Reactor() {}
+    void OnStarted(ServerContext* context) override {
+      ctx_ = context;
+      server_write_last_ = GetIntValueFromMetadata(
+          kServerFinishAfterNReads, context->client_metadata(), 0);
+      message_size_ = GetIntValueFromMetadata(kServerMessageSize,
+                                              context->client_metadata(), 0);
+      StartRead(&request_);
+    }
+    void OnDone() override {
+      GPR_ASSERT(finished_);
+      GPR_ASSERT(num_msgs_read_ == server_write_last_);
+      delete this;
+    }
+    void OnCancel() override {}
+    void OnReadDone(bool ok) override {
+      if (!ok) {
+        gpr_log(GPR_ERROR, "Server read failed");
```

Comment:
I would prefer going a different route and using !ok as a signal that the test is over rather than counting the number of reads. I'd prefer seeing num_msg_reads eliminated altogether.

Reply:
What's your reasoning?

Reply:
If I reuse !ok as a signal of streaming is over, would that be conflict with other error case, such as connection failed?

Reply:
nanahpang@ brings up a very reasonable point.

Reply:
So this is an interesting API hole that it looks like we've put ourselves into. In general.... the primary meaning of `!ok` for OnReadDone is that the client has nothing more to write. This is normally the only way of knowing that since it isn't always the case (unlike this code) that the client-to-server path has a specifiable finite number of messages.

The client would know if this was because of a stream error since the client would then get an error status returned regardless of what the server tries to send as a Status. So how does the server figure it out? If the server does a write and the OnWriteDone gets a `!ok` there was a stream failure. If the server sends a Finish and there's a stream error, we technically get the info internally saying that the finish completed with an error but we're swallowing that for now. So we actually never get the info at the server side that the stream had an error. The way around this is to have OnDone take a bool or something like that indicating whether the finish completed successfully.

Cc @sheenaqotj since we're discussing API implications

So I think the action items are
1. @nanahpang to make this code look like the rest of the tests and benchmarks by having the microbenchmark use `!ok` as a termination condition for a stream rather than a count passed by metadata
2. @vjpai to evaluate the API change needed to support the desired signalling

Reply:
Thanks for explaining!